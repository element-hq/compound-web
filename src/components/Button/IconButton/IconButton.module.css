/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.icon-button {
  --cpd-icon-button-indicator-border-size: calc(
    var(--cpd-icon-button-size) * 0.0625
  );

  inline-size: var(--cpd-icon-button-size);
  block-size: var(--cpd-icon-button-size);
  padding: calc(var(--cpd-icon-button-size) * 0.125);
  aspect-ratio: 1 / 1;
  color: var(--cpd-color-icon-tertiary);
  border: 0;
  appearance: none;
  cursor: pointer;
  border-radius: 50%;
  position: relative;
  background: transparent;
  line-height: 0;
}

.icon-button[aria-disabled="true"] {
  color: var(--cpd-color-icon-disabled);
  cursor: not-allowed;
}

.icon-button svg {
  inline-size: 100%;
  block-size: 100%;
}

.icon-button[data-indicator] svg {
  mask-image: url("./indicator-mask.svg");
  mask-position: center center;
  mask-repeat: no-repeat;
  mask-size: calc(var(--cpd-icon-button-size) * 0.75);
}

.icon-button[data-indicator]::before {
  content: "";
  position: absolute;
  inset-block-start: var(--cpd-icon-button-indicator-border-size);
  inset-inline-end: var(--cpd-icon-button-indicator-border-size);
  inline-size: calc(var(--cpd-icon-button-size) * 0.25);
  block-size: calc(var(--cpd-icon-button-size) * 0.25);
  border-radius: 50%;
  background: currentcolor;
}

.icon-button[data-indicator="highlight"]::before {
  background: var(--cpd-color-icon-success-primary);
}

/**
 * Hover state
 */

@media (hover) {
  .icon-button:not([aria-disabled="true"]):hover {
    color: var(--cpd-color-icon-primary);
    background: var(--cpd-color-bg-subtle-primary);
  }
}

.icon-button:not([aria-disabled="true"]):active {
  color: var(--cpd-color-icon-primary);
  background: var(--cpd-color-bg-subtle-primary);
}

@media (hover) {
  .icon-button:not([aria-disabled="true"])[data-indicator]:is(:hover)::before {
    /* Same colour as the background */
    border: var(--cpd-icon-button-indicator-border-size) solid
      var(--cpd-color-bg-subtle-primary);
    inset-block-start: 0;
    inset-inline-end: 0;
  }
}

.icon-button:not([aria-disabled="true"])[data-indicator]:is(:active)::before {
  /* Same colour as the background */
  border: var(--cpd-icon-button-indicator-border-size) solid
    var(--cpd-color-bg-subtle-primary);
  inset-block-start: 0;
  inset-inline-end: 0;
}
