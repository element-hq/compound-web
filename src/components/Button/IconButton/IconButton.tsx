/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { PropsWithChildren, forwardRef } from "react";
import classnames from "classnames";

import styles from "./IconButton.module.css";
import { UnstyledButton } from "../UnstyledButton";
import { UnstyledButtonPropsFor } from "../UnstyledButton";

type IconButtonProps = UnstyledButtonPropsFor<"button"> &
  JSX.IntrinsicElements["button"] & {
    /**
     * The CSS class name.
     */
    className?: string;
    /**
     * The avatar size in CSS units, e.g. `"24px"`.
     * @default 32px
     */
    size?: CSSStyleDeclaration["height"];
    /**
     * The icon button indicator displayed on the top right
     */
    indicator?: "default" | "highlight";
  };

/**
 * Display an icon as a button. Can render an indicator
 */
export const IconButton = forwardRef<
  HTMLButtonElement,
  PropsWithChildren<IconButtonProps>
>(function IconButton(
  { children, className, indicator, size = "32px", style, ...props },
  ref,
) {
  const classes = classnames(styles["icon-button"], className);
  return (
    <UnstyledButton
      as="button"
      ref={ref}
      className={classes}
      style={
        {
          "--cpd-icon-button-size": size,
          ...style,
        } as React.CSSProperties
      }
      {...props}
      data-indicator={indicator}
    >
      {React.Children.only(children)}
    </UnstyledButton>
  );
});
