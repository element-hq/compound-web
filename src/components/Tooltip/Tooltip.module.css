/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

.tooltip {
  font: var(--cpd-font-body-xs-medium);
  padding: var(--cpd-space-1-5x) var(--cpd-space-3x);
  background: var(--cpd-color-alpha-gray-1400);
  color: var(--cpd-color-text-on-solid-primary);
  border-radius: 4px;
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
}

.shortcut {
  font-weight: var(--cpd-font-weight-regular);
  color: var(--cpd-color-text-secondary);
}

.arrow {
  /* same color as the tooltip background */
  fill: var(--cpd-color-alpha-gray-1400);
}
