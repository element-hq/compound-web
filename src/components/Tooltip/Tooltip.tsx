/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React, { PropsWithChildren } from "react";
import {
  Root,
  Trigger,
  Portal,
  Content,
  Arrow,
  Provider,
} from "@radix-ui/react-tooltip";

import styles from "./Tooltip.module.css";
import classNames from "classnames";

type TooltipProps = {
  /**
   * The tooltip label
   */
  label: string;
  /**
   * The tooltip caption
   */
  caption?: string;
  /**
   * @deprecated replaced by `caption`
   */
  shortcut?: string;
  /**
   * The side where the tooltip is rendered
   * @default bottom
   */
  side?: React.ComponentProps<typeof Content>["side"];
  /**
   * The preferred alignment against the trigger.
   * May change when collisions occur.
   * @default center
   */
  align?: React.ComponentProps<typeof Content>["align"];
  /**
   * Event handler called when the escape key is down.
   */
  onEscapeKeyDown?: React.ComponentProps<typeof Content>["onEscapeKeyDown"];
  /**
   * Event handler called when a pointer event occurs outside
   * the bounds of the component.
   */
  onPointerDownOutside?: React.ComponentProps<
    typeof Content
  >["onPointerDownOutside"];
  /**
   * The controlled open state of the tooltip.
   * You will mostly want to omit this property. Will be used the vast majority
   * of the time during development.
   * @default undefined
   */
  open?: boolean;
};

/**
 * A tooltip component
 */
export const Tooltip = ({
  children,
  label,
  caption,
  shortcut,
  side = "bottom",
  align = "center",
  onEscapeKeyDown,
  onPointerDownOutside,
  open,
}: PropsWithChildren<TooltipProps>): JSX.Element => {
  return (
    <Provider>
      <Root open={open} delayDuration={300}>
        <Trigger asChild>{children}</Trigger>
        <Portal>
          <Content
            side={side}
            align={align}
            onEscapeKeyDown={onEscapeKeyDown}
            onPointerDownOutside={onPointerDownOutside}
            className={styles.tooltip}
          >
            {label}
            {/* Forcing dark theme, so that we have the correct contrast when
                using the text color secondary on a solid dark background. 
                This is temporary and should only remain until we figure out 
                the approach to on-solid tokens */}
            {(caption || shortcut) && (
              <span className={classNames(styles.caption, "cpd-theme-dark")}>
                {caption ?? shortcut}
              </span>
            )}
            <Arrow width={10} height={6} className={styles.arrow} />
          </Content>
        </Portal>
      </Root>
    </Provider>
  );
};
