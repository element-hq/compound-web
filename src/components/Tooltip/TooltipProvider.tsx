import { FloatingDelayGroup } from "@floating-ui/react";
import React, { type FC } from "react";

export const hoverDelay = { open: 300, close: 0 };

interface Props {
  children: React.ReactNode;
}

/**
 * Enables tooltips to share a global "warm-up" period for their hover delay.
 * You must wrap your application in this component for tooltips to function.
 */
export const TooltipProvider: FC<Props> = ({ children }) => (
  <FloatingDelayGroup delay={hoverDelay} timeoutMs={300}>
    {children}
  </FloatingDelayGroup>
);

TooltipProvider.displayName = "TooltipProvider";
