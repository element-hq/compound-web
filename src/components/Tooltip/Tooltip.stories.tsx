/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import React from "react";
import { Meta, StoryFn } from "@storybook/react";

import { Tooltip as TooltipComponent } from "./Tooltip";
import { IconButton } from "../Button";

import UserIcon from "@vector-im/compound-design-tokens/icons/user-profile.svg";

export default {
  title: "Tooltip",
  component: TooltipComponent,
  tags: ["autodocs"],
  argTypes: {},
  args: {},
} as Meta<typeof TooltipComponent>;

const TemplateSide: StoryFn<typeof TooltipComponent> = () => (
  <div
    style={{
      display: "flex",
      gap: "50px",
      flexDirection: "column",
      alignItems: "center",
    }}
  >
    <TooltipComponent open={true} side="top" label="@bob:example.org">
      <IconButton>
        <UserIcon />
      </IconButton>
    </TooltipComponent>
    <TooltipComponent open={true} side="right" label="@bob:example.org">
      <IconButton>
        <UserIcon />
      </IconButton>
    </TooltipComponent>
    <TooltipComponent open={true} side="bottom" label="@bob:example.org">
      <IconButton>
        <UserIcon />
      </IconButton>
    </TooltipComponent>
    <TooltipComponent open={true} side="left" label="@bob:example.org">
      <IconButton>
        <UserIcon />
      </IconButton>
    </TooltipComponent>
  </div>
);

export const Side = TemplateSide.bind({});
Side.args = {};

const TemplateAlign: StoryFn<typeof TooltipComponent> = () => (
  <div
    style={{
      display: "flex",
      gap: "50px",
      flexDirection: "column",
      alignItems: "center",
    }}
  >
    <TooltipComponent open={true} align="center" label="Copy" caption="⌘ + C">
      <IconButton>
        <UserIcon />
      </IconButton>
    </TooltipComponent>
    <TooltipComponent
      open={true}
      align="start"
      label="@bob:example.org"
      caption="⌘ + C"
    >
      <IconButton>
        <UserIcon />
      </IconButton>
    </TooltipComponent>
    <TooltipComponent
      open={true}
      align="end"
      label="@bob:example.org"
      caption="⌘ + C"
    >
      <IconButton>
        <UserIcon />
      </IconButton>
    </TooltipComponent>
  </div>
);

export const Align = TemplateAlign.bind({});
Align.args = {};
