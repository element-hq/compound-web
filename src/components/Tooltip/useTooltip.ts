/*
 *
 * Copyright 2024 New Vector Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 * /
 */

import {
  arrow,
  autoUpdate,
  flip,
  offset,
  OpenChangeReason,
  Placement,
  shift,
  useDelayGroup,
  useDismiss,
  useFloating,
  useFocus,
  useHover,
  useId,
  useInteractions,
  useRole,
} from "@floating-ui/react";
import {
  useMemo,
  useRef,
  useState,
  JSX,
  AriaAttributes,
  useEffect,
} from "react";
import { hoverDelay } from "./TooltipProvider";

export interface CommonUseTooltipProps {
  /**
   * The controlled open state of the tooltip.
   * If provided, the tooltip will be in controlled mode.
   * When true, the tooltip is always open. When false, the tooltip is always hidden.
   * When undefined, the tooltip will manage its own open state.
   * You will mostly want to omit this property. Will be used the vast majority
   * of the time during development.
   * @default undefined
   */
  open?: boolean;
  /**
   * Whether the tooltip should be forced to be in a closed state.
   */
  // TODO: Deprecate this? It seems redundant to open: false.
  disabled?: boolean;
  /**
   * The caption of the tooltip.
   * JSX.Element can be used to provide accessibility content like kbd element.
   * Keep in mind, the caption should not be used for interactive content.
   */
  caption?: string | JSX.Element;
  /**
   * The event handler for the open change.
   */
  onOpenChange?: (
    open: boolean,
    event?: Event | undefined,
    reason?: OpenChangeReason | undefined,
  ) => void;
  /**
   * The placement of the tooltip.
   * @default "bottom"
   */
  placement?: Placement;
  /**
   * Whether the trigger element is interactive.
   * When trigger is interactive:
   *      - tooltip will be shown after a 300ms delay.
   * When trigger is not interactive:
   *      - tooltip will be shown instantly when pointer enters trigger.
   *      - trigger will be wrapped in a span with a tab index from prop nonInteractiveTriggerTabIndex
   */
  isTriggerInteractive: boolean;

  /**
   * Additional aria-* attributes to pass through to the floating tooltip for
   * edge cases which require more user awareness like errors & alerts.
   */
  "aria-atomic"?: AriaAttributes["aria-atomic"];
  "aria-live"?: AriaAttributes["aria-live"];
}

export interface TooltipLabel {
  /**
   * A label for the target element.
   */
  label: string;
}

export interface TooltipDescription {
  /**
   * A description for the target element.
   */
  description: string;
}

type UseTooltipProps = CommonUseTooltipProps &
  (TooltipLabel | TooltipDescription);

export function useTooltip({
  open: controlledOpen,
  disabled = false,
  onOpenChange,
  placement = "bottom",
  isTriggerInteractive,
  caption,
  "aria-atomic": ariaAtomic,
  "aria-live": ariaLive,
  ...props
}: UseTooltipProps) {
  const labelId = useId();
  const captionId = useId();
  const arrowRef = useRef(null);

  const [uncontrolledOpen, setUncontrolledOpen] = useState(false);

  // Use controlledOpen if it is provided, otherwise use uncontrolledOpen
  const open = disabled ? false : (controlledOpen ?? uncontrolledOpen);
  const setOpen = (
    open: boolean,
    event?: Event | undefined,
    reason?: OpenChangeReason | undefined,
  ) => {
    onOpenChange?.(open, event, reason);
    // we are in uncontrolled mode
    if (controlledOpen === undefined) setUncontrolledOpen(open);
  };

  const data = useFloating({
    placement,
    open,
    onOpenChange: setOpen,
    whileElementsMounted: autoUpdate,
    middleware: [
      // arrow height 6px
      offset(6),
      flip({
        crossAxis: placement.includes("-"),
        fallbackAxisSideDirection: "start",
        padding: 5,
      }),
      shift({ padding: 5 }),
      // add the little arrow along with the floating content
      arrow({
        element: arrowRef,
      }),
    ],
  });

  const context = data.context;
  const { delay, initialDelay } = useDelayGroup(context);
  // We can tell if no delay group has been provided, because the delay will
  // default to zero
  if (initialDelay !== hoverDelay)
    throw new Error("Tooltips must be wrapped in a global <TooltipProvider>");

  const hover = useHover(context, {
    move: false,
    enabled: controlledOpen === undefined,
    // Show tooltip after a delay when trigger is interactive
    delay: isTriggerInteractive ? delay : {},
    mouseOnly: true,
  });

  const focus = useFocus(context, {
    enabled: controlledOpen === undefined,
  });

  // On touch screens, show the tooltip on a long press
  const pressTimer = useRef<number | undefined>(undefined);
  useEffect(() => () => window.clearTimeout(pressTimer.current), []);
  const press = useMemo(() => {
    const onTouchEnd = () => {
      if (pressTimer.current === undefined)
        pressTimer.current = window.setTimeout(() => {
          setOpen(false);
          pressTimer.current = undefined;
        }, 1500);
      else window.clearTimeout(pressTimer.current);
    };
    return {
      // Set these props on the anchor element
      reference: {
        onTouchStart: () => {
          if (pressTimer.current !== undefined)
            window.clearTimeout(pressTimer.current);
          pressTimer.current = window.setTimeout(() => {
            setOpen(true);
            pressTimer.current = undefined;
          }, 500);
        },
        onTouchEnd,
        onTouchCancel: onTouchEnd,
      },
    };
  }, []);

  const dismiss = useDismiss(context);

  const purpose = "label" in props ? "label" : "description";
  // A descriptive tooltip should set role="tooltip" and aria-describedby
  const role = useRole(context, {
    enabled: purpose === "description",
    role: "tooltip",
  });

  // A label tooltip should set aria-labelledby with no role regardless of
  // whether the tooltip is visible.
  // (Source: https://zoebijl.github.io/apg-tooltip/#tooltip-main-label)
  // useRole doesn't support this use case correctly, so we do it manually.
  const label = useMemo(
    () =>
      purpose === "label"
        ? {
            // Set these props on the anchor element
            reference: {
              "aria-labelledby": labelId,
              "aria-describedby": caption ? captionId : undefined,
            },
          }
        : {},
    [purpose, labelId, captionId],
  );

  const interactions = useInteractions([
    hover,
    focus,
    press,
    dismiss,
    role,
    label,
  ]);

  return useMemo(
    () => ({
      labelId,
      captionId: caption ? captionId : undefined,
      caption,
      purpose: purpose as "label" | "description",
      open,
      setOpen,
      tooltipProps: {
        "aria-atomic": ariaAtomic,
        "aria-live": ariaLive,
      },
      ...interactions,
      ...data,
      arrowRef,
    }),
    [
      labelId,
      captionId,
      caption,
      role,
      open,
      setOpen,
      interactions,
      data,
      arrowRef,
    ],
  );
}
