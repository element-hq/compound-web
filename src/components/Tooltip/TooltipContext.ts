/*
 * Copyright 2024 New Vector Ltd.
 *
 * SPDX-License-Identifier: AGPL-3.0-only OR LicenseRef-Element-Commercial
 * Please see LICENSE files in the repository root for full details.
 */

import { createContext, useContext } from "react";
import { useTooltip } from "./useTooltip";

type ContextType = ReturnType<typeof useTooltip> | null;
/**
 * The context for the Tooltip components.
 */
export const TooltipContext = createContext<ContextType>(null);

/**
 * Provides the context for the Tooltip components.
 */
export function useTooltipContext() {
  const context = useContext(TooltipContext);

  if (context == null) {
    throw new Error("Tooltip components must be wrapped in <Tooltip />");
  }

  return context;
}
