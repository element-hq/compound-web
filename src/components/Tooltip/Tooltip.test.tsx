/*
Copyright 2023 New Vector Ltd

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { describe, it, expect, beforeAll } from "vitest";
import { render } from "@testing-library/react";
import React from "react";

import { Tooltip } from "./Tooltip";
import { IconButton } from "../IconButton/IconButton";

describe("Tooltip", () => {
  beforeAll(() => {
    global.ResizeObserver = require("resize-observer-polyfill");
  });
  it("renders open by default", () => {
    const { asFragment } = render(
      <Tooltip label="Hello world ðŸ‘‹" caption="âŒ˜ + C" open={true}>
        <IconButton>
          <svg />
        </IconButton>
      </Tooltip>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
  it("renders", () => {
    const { asFragment } = render(
      <Tooltip label="Hello world ðŸ‘‹">
        <IconButton>
          <svg />
        </IconButton>
      </Tooltip>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
