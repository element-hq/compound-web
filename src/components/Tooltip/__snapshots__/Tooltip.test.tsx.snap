// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Tooltip > renders 1`] = `
<DocumentFragment>
  <button
    class="_icon-button_e3253e"
    data-state="closed"
    role="button"
    style="--cpd-icon-button-size: 32px;"
    tabindex="0"
  >
    <svg />
  </button>
</DocumentFragment>
`;

exports[`Tooltip > renders open by default 1`] = `
<DocumentFragment>
  <button
    aria-describedby="radix-:r0:"
    class="_icon-button_e3253e"
    data-state="instant-open"
    role="button"
    style="--cpd-icon-button-size: 32px;"
    tabindex="0"
  >
    <svg />
  </button>
</DocumentFragment>
`;
